server/game_components/game_objects.py

```diff
@@ -0,0 +1,533 @@
+        if _target is None:
+            assert self.requires_target is False
+        """
+        Perform an action
+
+        This is the combat of the game. It can handle single and multiple targets.
+        correct usage of number of targets is NOT taken into account.
+        This "deals damage" to the target if it hits and already takes into account cost
+        for each element in the list:
+            hit - did the action hit the target
+            cast - was the spell cast or not
+            dmg - how much dmg would have occurred or did occur.
+            caster - the caster uid
+            target - the target uid
+        FOR EXAMPLE:
+            hit - false
+            dmg - 1000
+            cast - true
+            caster - the caster uid
+            target - the target uid
+            the spell did cast but did not hit, it WOULD HAVE done 1000 dmg BUT DIDN'T
+
+        :param _caster: the caster
+        :param _target: the target(s)
+        :return: its a list of what happened. Single target attacks still return a list of dicts
+        """
+        hit = False
+        cast = False
+        if _caster.mana >= self.__cost:
+            _caster.mana -= self.__cost
+            cast = True
+
+        if _target is None:
+            if self.area_of_effect:
+                action_list = []
+                targets = []
+                for mob in _caster.in_room.get_mobs():
+                    targets.append(mob)
+                for player in _caster.in_room.get_players():
+                    if player.uid != _caster.uid:
+                        targets.append(player)
+                for entity in targets:
+                    dmg = random.randint(self.__min_damage, self.__max_damage)
+                    if cast:
+                        hit_check = random.randint(0, 100)
+                        if hit_check <= self.__hit_percentage:
+                            hit = True
+                        if hit:
+                            entity.health -= dmg
+                        action_list.append(
+                            {
+                                "caster": _caster.uid,
+                                "target": entity.uid,
+                                "hit": hit,
+                                "dmg": dmg,
+                                "cast": cast,
+                            }
+                        )
+                    else:
+                        action_list.append(
+                            {
+                                "caster": _caster.uid,
+                                "target": entity.uid,
+                                "hit": hit,
+                                "dmg": dmg,
+                                "cast": cast,
+                            }
+                        )
+                return action_list
+            else:
+                _target = _caster
+
+        if isinstance(_target, Entity) and (not self.area_of_effect):
```

Comment:
The parenthesis around `not self.area_of_effect` is unnecessary, you can omit them.